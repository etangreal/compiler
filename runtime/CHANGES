7 Dec 2012

Added get_fun_ptr_from_attr. See a use of it in test4.c.

Added a call to a method with two arguments to test4.c.

create_xxx and inject_xxx can now both be used with boxing or tagging runtime
implementations. create_xxx now calls inject_xxx, which does boxing or tagging
depending on if RTFLAGS=-DTAGGING is passed to make.

Changed project_int to coerce bools and floats to ints.
Changed project_bool to coerce ints and floats to bools.
Changed project_float to coerce ints and bools to floats.
