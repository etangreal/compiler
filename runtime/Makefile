# RTFLAGS = 
# RTFLAGS = -DTAGGING

CC = gcc
CFLAGS = -m32 $(RTFLAGS)

# debugging support
# CFLAGS = -g -m32 $(RTFLAGS)

RUNTIME = hashtable.o hashtable_itr.o hashtable_utility.o runtime.o

#GC = lib/libgc.a

runtime: $(RUNTIME)
	@true

test: test1 test2 test3 test4 test5

return: $(RUNTIME) return.o
	@$(CC) $(CFLAGS) -S return.c -o return.s $(GC)
	@$(CC) $(CFLAGS) return.s $(RUNTIME) -o return $(GC)

last: $(RUNTIME) last.o
	@$(CC) $(CFLAGS) -S last.c -o last.s $(GC)
	@$(CC) $(CFLAGS) last.s $(RUNTIME) -o last $(GC)

t2: $(RUNTIME) t2.o
	@$(CC) $(CFLAGS) -S t2.c -o t2.s $(GC)
	@$(CC) $(CFLAGS) t2.s $(RUNTIME) -o t2 $(GC)

t10: $(RUNTIME) t0.o
	@$(CC) $(CFLAGS) -S t10.c -o t10.s $(GC)
	@$(CC) $(CFLAGS) t10.s $(RUNTIME) -o t10 $(GC)

t1: $(RUNTIME) t1.o
	@$(CC) $(CFLAGS) -S t1.c -o t1.s $(GC)
	@$(CC) $(CFLAGS) t1.s $(RUNTIME) -o t1 $(GC)
	
t: $(RUNTIME) t.o
	@$(CC) $(CFLAGS) -S t.c -o t.s $(GC)
	@$(CC) $(CFLAGS) t.s $(RUNTIME) -o t $(GC)

tt: $(RUNTIME) tt.o
	@$(CC) $(CFLAGS) -S tt.c -o tt.s $(GC)
	@$(CC) $(CFLAGS) tt.s $(RUNTIME) -o tt $(GC)

test0: $(RUNTIME) test0.o
	@$(CC) $(CFLAGS) test0.s $(RUNTIME) -o test0 $(GC)

test1: $(RUNTIME) test1.o
	@$(CC) $(CFLAGS) -o test1 $(RUNTIME) test1.o $(GC)
	@./test1 > test1.out
	@if cmp test1.out test1.expected; then echo test1 ok ; else echo test1 failed ; fi

test2: $(RUNTIME) test2.o
	@$(CC) $(CFLAGS) -o test2 $(RUNTIME) test2.o $(GC)
	@echo 1 | ./test2 > test2.out
	@if cmp test2.out test2.expected; then echo test2 ok ; else echo test2 failed ; fi

test2.1: $(RUNTIME) test2.1.o
	#gcc -m32 -S test1.c -o test1.s
	#gcc -m32 test1.s runtime.c -o test1
	@$(CC) $(CFLAGS) -S test2.1.c -o test2.1.s $(GC)
	@$(CC) $(CFLAGS) test2.1.s $(RUNTIME) -o test2.1 $(GC)
	@echo 1 | ./test2.1 > test2.1.out
	@if cmp test2.1.out test2.1.expected; then echo test2.1 ok ; else echo test2.1 failed ; fi

test3: $(RUNTIME) test3.o
	@$(CC) $(CFLAGS) -o test3 $(RUNTIME) test3.o $(GC)
	@echo 1 | ./test3 > test3.out
	@if cmp test3.out test3.expected; then echo test3 ok ; else echo test3 failed ; fi

test4: $(RUNTIME) test4.o
	@$(CC) $(CFLAGS) -o test4 $(RUNTIME) test4.o $(GC)
	@./test4 > test4.out
	@if cmp test4.out test4.expected; then echo test4 ok ; else echo test4 failed ; fi

test5: $(RUNTIME) test5.o
	@$(CC) $(CFLAGS) -o test5 $(RUNTIME) test5.o $(GC)
	@./test5 > test5.out
	@if cmp test5.out test5.expected; then echo test5 ok ; else echo test5 failed ; fi

retest: test-clean test

test-clean:
	-rm -f test*.o
	-rm -f test*.out
	-rm -f test[0-9] test[0-9][0-9] test[0-9][0-9][0-9] test[0-9]\.[0-9]

clean: test-clean
	-rm -f hashtable.o hashtable_itr.o hashtable_utility.o runtime.o
	-rm -rf *.dSYM
	-rm -f *.s

.c.o:
	$(CC) $(CFLAGS) -c $< -o $@
