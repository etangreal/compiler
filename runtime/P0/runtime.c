#include <stdio.h>

void print_int(int x)
{
    printf("%d", x);
}

void print_int_nl(int x)
{
    printf("%d\n", x);
}

int input()
{
    int x;
   
    scanf("%d", &x);
    return x;
}
