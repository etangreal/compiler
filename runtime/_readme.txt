ssh atelier.inf.usi.ch -l salzmane

scp /path/to/local/file username@hostname:/path/to/remote/file
scp /Users/ernst/Dropbox/code/compilers/compiler/* salzmane@atelier.inf.usi.ch:~/compiler
scp -r /Users/ernst/Dropbox/code/compilers/compiler/* salzmane@atelier.inf.usi.ch:~/compiler

gcc -m32 -S test1.c -o test1.s
gcc -m32 test1.s runtime.c -o test1
