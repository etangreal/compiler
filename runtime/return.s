	.section	__TEXT,__text,regular,pure_instructions
	.globl	_main
	.align	4, 0x90
_main:
	pushl	%ebp
	movl	%esp, %ebp
	subl	$56, %esp
	call	L1$pb
L1$pb:
	popl	%eax
	movl	%eax, -40(%ebp)
	call	_setup_classes
	movl	-40(%ebp), %eax
	movl	L_CPoint$non_lazy_ptr-L1$pb(%eax), %ecx
	movl	(%ecx), %ecx
	movl	%ecx, (%esp)
	call	_create_object
	movl	%eax, -12(%ebp)
	movl	-12(%ebp), %eax
	movl	-40(%ebp), %ecx
	leal	L_.str-L1$pb(%ecx), %edx
	movl	%eax, (%esp)
	movl	%edx, 4(%esp)
	call	_get_fun_ptr_from_attr
	movl	%eax, -20(%ebp)
	movl	$2, (%esp)
	call	_create_int
	movl	-20(%ebp), %ecx
	movl	-12(%ebp), %edx
	movl	%edx, (%esp)
	movl	%eax, 4(%esp)
	call	*%ecx
	movl	-12(%ebp), %eax
	movl	-40(%ebp), %ecx
	leal	L_.str1-L1$pb(%ecx), %ecx
	movl	%eax, (%esp)
	movl	%ecx, 4(%esp)
	call	_get_attr
	movl	%eax, -24(%ebp)
	movl	-24(%ebp), %eax
	movl	%eax, (%esp)
	call	_get_function
	movl	%eax, -28(%ebp)
	movl	-28(%ebp), %eax
	movl	%eax, (%esp)
	call	_get_fun_ptr
	movl	%eax, -32(%ebp)
	movl	-32(%ebp), %eax
	movl	%eax, -36(%ebp)
	movl	-24(%ebp), %eax
	movl	%eax, (%esp)
	call	_get_receiver
	movl	-36(%ebp), %ecx
	movl	%eax, (%esp)
	call	*%ecx
	movl	%eax, -16(%ebp)
	movl	-16(%ebp), %eax
	movl	%eax, (%esp)
	call	_print_any
	movl	$0, -8(%ebp)
	movl	-8(%ebp), %eax
	movl	%eax, -4(%ebp)
	movl	-4(%ebp), %eax
	addl	$56, %esp
	popl	%ebp
	ret

	.align	4, 0x90
_setup_classes:
	pushl	%ebp
	movl	%esp, %ebp
	pushl	%edi
	pushl	%esi
	subl	$48, %esp
	call	L2$pb
L2$pb:
	popl	%eax
	movl	$0, (%esp)
	movl	%eax, -32(%ebp)
	call	_create_int
	movl	%eax, -12(%ebp)
	movl	$1, (%esp)
	call	_create_int
	movl	%eax, -16(%ebp)
	movl	-12(%ebp), %eax
	movl	%eax, (%esp)
	call	_create_list
	movl	%eax, -20(%ebp)
	movl	-20(%ebp), %eax
	movl	%eax, (%esp)
	call	_create_class
	movl	-32(%ebp), %ecx
	movl	L_CPoint$non_lazy_ptr-L2$pb(%ecx), %edx
	movl	%eax, (%edx)
	movl	-20(%ebp), %eax
	leal	_CPoint___init___-L2$pb(%ecx), %esi
	movl	%esi, (%esp)
	movl	%eax, 4(%esp)
	movl	%edx, -36(%ebp)
	call	_create_closure
	movl	%eax, -24(%ebp)
	movl	-36(%ebp), %eax
	movl	(%eax), %ecx
	movl	-24(%ebp), %edx
	movl	-32(%ebp), %esi
	leal	L_.str-L2$pb(%esi), %edi
	movl	%ecx, (%esp)
	movl	%edi, 4(%esp)
	movl	%edx, 8(%esp)
	call	_set_attr
	movl	-20(%ebp), %eax
	leal	_CPoint_getX_-L2$pb(%esi), %ecx
	movl	%ecx, (%esp)
	movl	%eax, 4(%esp)
	call	_create_closure
	movl	%eax, -28(%ebp)
	movl	-36(%ebp), %eax
	movl	(%eax), %eax
	movl	-28(%ebp), %ecx
	leal	L_.str1-L2$pb(%esi), %edx
	movl	%eax, (%esp)
	movl	%edx, 4(%esp)
	movl	%ecx, 8(%esp)
	call	_set_attr
	addl	$48, %esp
	popl	%esi
	popl	%edi
	popl	%ebp
	ret

	.globl	_CPoint___init___
	.align	4, 0x90
_CPoint___init___:
	pushl	%ebp
	movl	%esp, %ebp
	subl	$24, %esp
	call	L3$pb
L3$pb:
	popl	%eax
	movl	12(%ebp), %ecx
	movl	8(%ebp), %edx
	movl	%edx, -4(%ebp)
	movl	%ecx, -8(%ebp)
	movl	-4(%ebp), %ecx
	movl	-8(%ebp), %edx
	leal	L_.str2-L3$pb(%eax), %eax
	movl	%ecx, (%esp)
	movl	%eax, 4(%esp)
	movl	%edx, 8(%esp)
	call	_set_attr
	movl	-12(%ebp), %eax
	addl	$24, %esp
	popl	%ebp
	ret

	.globl	_CPoint_getX_
	.align	4, 0x90
_CPoint_getX_:
	pushl	%ebp
	movl	%esp, %ebp
	subl	$24, %esp
	movl	8(%ebp), %eax
	movl	%eax, -4(%ebp)
	movl	$200, (%esp)
	call	_create_int
	movl	%eax, -16(%ebp)
	movl	-16(%ebp), %eax
	movl	%eax, -12(%ebp)
	movl	-12(%ebp), %eax
	movl	%eax, -8(%ebp)
	movl	-8(%ebp), %eax
	addl	$24, %esp
	popl	%ebp
	ret

	.comm	_CPoint,4,2
	.section	__TEXT,__cstring,cstring_literals
L_.str:
	.asciz	 "__init__"

L_.str1:
	.asciz	 "getX"

L_.str2:
	.asciz	 "x_"


	.section	__IMPORT,__pointers,non_lazy_symbol_pointers
L_CPoint$non_lazy_ptr:
.indirect_symbol _CPoint
	.long	0

.subsections_via_symbols
