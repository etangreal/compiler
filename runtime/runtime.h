#ifndef RUNTIME_H
#define RUNTIME_H

#include "hashtable.h"
#include "hashtable_itr.h"
#include "hashtable_utility.h"

/* for old times sake */
void print_int_nl(int x);
int input();

/* Structure and type-tag definitions   */

#ifdef TAGGING

#define MASK 3    /* 11 */
#define SHIFT 2

#define INT_TAG 0     /* 00 */

#define BOOL_TAG 1   /* 01 */

/* The following is for a non-standard single-precision (21-bit fraction)
   floating point number */
#define FLOAT_TAG 2   /* 10 */

/*
  For larger objects, we have a pointer to the object, and the
  object starts with a tag that says what it is.
*/
#define BIG_TAG 3   /* 11 */
#endif

enum big_type_tag {
  LIST, DICT, FUN, CLASS, OBJECT, UBMETHOD, BMETHOD,
  INT, FLOAT, BOOL
};

struct pyobj_struct;

#ifdef TAGGING
typedef long int pyobj;
#else
typedef struct pyobj_struct *pyobj;
#endif

struct list_struct {
  pyobj* data;
  unsigned int len;
};
typedef struct list_struct list;

typedef struct hashtable* dict;

struct fun_struct {
  void* function_ptr;
  pyobj free_vars;
};
typedef struct fun_struct function;

struct class_struct {
  struct hashtable *attrs;
  int nparents;
  struct class_struct *parents;
};
typedef struct class_struct class;

struct object_struct {
  struct hashtable *attrs;
  class cl;
};
typedef struct object_struct object;

struct unbound_method_struct {
  function fun;
  class cl;
};
typedef struct unbound_method_struct unbound_method;

struct bound_method_struct {
  function fun;
  object receiver;
};
typedef struct bound_method_struct bound_method;

struct pyobj_struct {
  enum big_type_tag tag;
  union {
#ifndef TAGGING
    int i;
    float fl;
    int b; /* boolean */
#endif
    dict d;
    list l;
    function f;
    class cl;
    object obj;
    unbound_method ubm;
    bound_method bm;
  } u;
};
typedef struct pyobj_struct big_pyobj;

int tag(pyobj val);

int is_int(pyobj val);
int is_bool(pyobj val);
int is_big(pyobj val);
int is_function(pyobj val);
int is_object(pyobj val);
int is_class(pyobj val);
int is_unbound_method(pyobj val);
int is_bound_method(pyobj val);

/* Boxing/tagging operations */
pyobj inject_int(int i);
pyobj inject_bool(int b);
pyobj inject_float(float f);

#ifdef TAGGING
pyobj inject_big(big_pyobj* p);
#endif

/* Boxing/tagging operations. Just calls inject_T functions */
pyobj create_int(int i);
pyobj create_bool(int b);
pyobj create_float(float f);

/* Unboxing/untagging operations */
int project_int(pyobj val);
int project_bool(pyobj val);
float project_float(pyobj val);
big_pyobj* project_big(pyobj val);

/* Operations */

int is_true(pyobj v);
void print_any(pyobj p);
pyobj input_int();

pyobj create_list(pyobj length);
pyobj create_dict();
pyobj set_subscript(pyobj c, pyobj key, pyobj val);
pyobj get_subscript(pyobj c, pyobj key);

pyobj append(pyobj a, pyobj b);
int equal(pyobj a, pyobj b);
int not_equal(pyobj x, pyobj y);

pyobj create_closure(void* fun_ptr, pyobj free_vars);
void* get_fun_ptr(pyobj);
pyobj get_free_vars(pyobj);
pyobj set_free_vars(pyobj b, pyobj free_vars);

void *get_fun_ptr_from_attr(pyobj c, char* attr);

pyobj create_class(pyobj bases); /* bases should be a list of classes */
pyobj create_class_without_superclass(); /* calls create_class with an empty list */
pyobj create_class_with_superclass(pyobj base); /* calls create_class with a singleton list */
pyobj create_object(pyobj cl);
int inherits(pyobj c1, pyobj c2); /* Returns true if class c1 inherits from class c2 */
pyobj get_class(pyobj o); /* Get the class from an object or unbound method */
pyobj get_receiver(pyobj o); /* Get the receiver from inside a bound method */
pyobj get_function(pyobj o); /* Get the function from inside a method */
int has_attr(pyobj o, char* attr);
pyobj get_attr(pyobj c, char* attr);
pyobj set_attr(pyobj obj, char* attr, pyobj val);

pyobj error_pyobj(char* string);

#endif /* RUNTIME_H */
