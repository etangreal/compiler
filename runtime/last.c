#include <stdlib.h>
#include <assert.h>
#include "runtime.h"

/* methods */
pyobj C_m(pyobj self);
pyobj D_m(pyobj self);
pyobj D_n(pyobj self, pyobj x);

/* classes */
pyobj C, D;

static void setup_classes();

int main()
{
  setup_classes();

  // c = C()
  pyobj c = create_object(C);

  // d = D()
  pyobj d = create_object(D);

  // TODO: call the __init__ method if it exists

#ifdef TAGGING
  pyobj one = inject_int(1);
  pyobj three = inject_int(3);
#else
  pyobj one = create_int(1);
  pyobj three = create_int(3);
#endif

  // c.f = 1
  set_attr(c, "f", one);

  // d.f = 1
  set_attr(d, "f", one);

  pyobj i, j, k, h;

  // i = c.m()
  {
    pyobj meth = get_attr(c, "m");
    pyobj fun = get_function(meth);
    void *fp = get_fun_ptr(fun);
    pyobj (*f)(pyobj) = (pyobj (*)(pyobj)) fp; // cast to a function pointer type
    i = f(get_receiver(meth));
  }

  // j = d.m()
  {
    pyobj meth = get_attr(d, "m");
    pyobj fun = get_function(meth);
    void *fp = get_fun_ptr(fun);
    pyobj (*f)(pyobj) = (pyobj (*)(pyobj)) fp; // cast to a function pointer type
    j = f(get_receiver(meth));
  }

  // d.n(3)
  {
    pyobj (*f)(pyobj, pyobj) = (pyobj (*)(pyobj, pyobj)) get_fun_ptr_from_attr(d, "n");
    f(d, three);
  }

  // k = d.m()
  {
    pyobj meth = get_attr(d, "m");
    pyobj fun = get_function(meth);
    void *fp = get_fun_ptr(fun);
    pyobj (*f)(pyobj) = (pyobj (*)(pyobj)) fp; // cast to a function pointer type
    k = f(get_receiver(meth));
  }

  // h = i + j + k
  {
#ifdef TAGGING
    // optimized, but assumes i and j are integers
    // h = i + j + k

    // unoptimized, but checks that i and j are integers
    h = inject_int(project_int(i) + project_int(j) + project_int(k));
#else
    h = create_int(project_int(i) + project_int(j) + project_int(k));
#endif
  }

  // print i, j, k
  print_any(i);
  print_any(j);
  print_any(k);
  print_any(h);
}

/* code to create classes C and D */
static void setup_classes()
{
#ifdef TAGGING
  pyobj zero = inject_int(0);
  pyobj one = inject_int(1);
#else
  pyobj zero = create_int(0);
  pyobj one = create_int(1);
#endif
  pyobj list0 = create_list(zero);
  pyobj list1 = create_list(one);

  // class C:
  //   def m(self):
  //     return self.f
  C = create_class(list0);

  // Add method m to the class.
  set_attr(C, "m", create_closure(C_m, list0));

  // class D(C):
  //   def m(self):
  //     return self.f + 1
  set_subscript(list1, zero, C); // list1[0] = C
  D = create_class(list1);

  pyobj D_m_closure = create_closure(D_m, list0);
  set_attr(D, "m", create_closure(D_m, list0));

  pyobj D_n_closure = create_closure(D_n, list0);
  set_attr(D, "n", create_closure(D_n, list0));
}

/*
 * def m(self): return self.f
 */
pyobj C_m(pyobj self) {
  return get_attr(self, "f");
}

/*
 * def m(self): return self.f + 1
 */
pyobj D_m(pyobj self) {
  pyobj f = get_attr(self, "f");
  int i = project_int(f);
#ifdef TAGGING
  return inject_int(i+1);
#else
  return create_int(i+1);
#endif
}

/*
 * def n(self, x): self.f = x; return None
 */
pyobj D_n(pyobj self, pyobj x) {
  set_attr(self, "f", x);
#ifdef TAGGING
  return inject_big(0); /* None */
#else
  return NULL;
#endif
}
