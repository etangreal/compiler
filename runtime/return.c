

// ------------------------------------------------------------------------------------------------

#include <stdlib.h>
#include <assert.h>
#include "runtime.h"

// ------------------------------------------------------------------------------------------------

/*
	class Point:

		x_ = 1

		def __init__ (self, x, y):
			self.x_ = x
			self.y_ = y

		def move(self,dx,dy):
			self.x_ = self.x_ + dx
			self.y_ = self.y_ + dy

		def getX(self):
			return self.x_

	p = Point(2,4)

	x = p.getX()

	if( p.x_ < p.y_ ):
		p.move(10,10)

	print p.x_ + p.y_
*/

// ------------------------------------------------------------------------------------------------

/* forward */

static void setup_classes();
pyobj CPoint___init___(pyobj self, pyobj x);

pyobj CPoint_getX_(pyobj self);

/* classes */
pyobj CPoint;

int main() {

	setup_classes();

	// p = CPoint()
	pyobj p = create_object(CPoint);



	// ----------------------------------------------

	// p.__init__(2)
	{
		pyobj (*f)(pyobj, pyobj) = (pyobj (*)(pyobj, pyobj)) get_fun_ptr_from_attr(p, "__init__");
		f(p, create_int(2));
	}


	// ----------------------------------------------

	pyobj x_;
	// x_ = p.getX()
	{
		pyobj meth = get_attr(p, "getX");
		pyobj fun = get_function(meth);
		void *fp = get_fun_ptr(fun);
		pyobj (*f)(pyobj) = (pyobj (*)(pyobj)) fp; // cast to a function pointer type
		x_ = f(get_receiver(meth));

		//print: 3
		print_any(x_);
	}

	// ----------------------------------------------

	return 0;
}

// ------------------------------------------------------------------------------------------------

static void setup_classes() {

	pyobj zero = create_int(0);
	pyobj one = create_int(1);

	pyobj list0 = create_list(zero);
	
	CPoint = create_class(list0);

	//create function: __init__
	pyobj CPoint___init___closure = create_closure(CPoint___init___, list0);
	set_attr(CPoint, "__init__", CPoint___init___closure);

	//create function: getX
	pyobj CPoint_getX_closure = create_closure(CPoint_getX_, list0);
	set_attr(CPoint, "getX", CPoint_getX_closure);
}

// ------------------------------------------------------------------------------------------------

/*
 * class Point:
 *
 * 		def __init__(self, x, y): 
 * 			self.x_ = x
 *			self.y_ = y
 *			return None
 */

pyobj CPoint___init___(pyobj self, pyobj x) {

	set_attr(self, "x_", x);

}

// ------------------------------------------------------------------------------------------------



// ------------------------------------------------------------------------------------------------

/*
 * class Point:
 *
 *		def getX(self):
 *			return self.x_
 *
 */

pyobj CPoint_getX_(pyobj self) {

	pyobj x;
	x = create_int(200);
	return x;
}

// ------------------------------------------------------------------------------------------------

