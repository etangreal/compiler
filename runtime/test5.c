#include <stdlib.h>
#include <assert.h>
#include "runtime.h"

/* classes */
pyobj C, D;

static void setup_classes();

int main()
{
  setup_classes();

  // c = C()
  pyobj c = create_object(C);

  // d = D()
  pyobj d = create_object(D);

  // TODO: call the __init__ method if it exists

#ifdef TAGGING
  pyobj one = inject_int(1);
  pyobj three = inject_int(3);
#else
  pyobj one = create_int(1);
  pyobj three = create_int(3);
#endif

  // c.f = 1
  set_attr(c, "f", one);

  // d.f = 1
  set_attr(d, "f", three);

  print_any(inject_bool(equal(one, one)));
  print_any(inject_bool(equal(c, c)));
  print_any(inject_bool(equal(get_attr(c, "f"), get_attr(c, "f"))));
  print_any(inject_bool(equal(inject_bool(0), inject_bool(0))));
  print_any(inject_bool(equal(inject_bool(1), inject_bool(1))));

  print_any(one);
  print_any(three);
  print_any(inject_bool(equal(one, three)));
  print_any(c);
  print_any(d);
  print_any(inject_bool(equal(c, d)));
  print_any(inject_bool(0));
  print_any(inject_bool(1));
  print_any(inject_bool(equal(inject_bool(0), inject_bool(1))));

  print_any(get_attr(c, "f"));
  print_any(get_attr(d, "f"));
  print_any(inject_bool(equal(get_attr(c, "f"), get_attr(d, "f"))));
}

/* code to create classes C and D */
static void setup_classes()
{
#ifdef TAGGING
  pyobj zero = inject_int(0);
  pyobj one = inject_int(1);
#else
  pyobj zero = create_int(0);
  pyobj one = create_int(1);
#endif
  pyobj list0 = create_list(zero);
  pyobj list1 = create_list(one);

  // class C:
  //   pass
  C = create_class(list0);

  // class D(C):
  //   pass
  set_subscript(list1, zero, C); // list1[0] = C
  D = create_class(list1);
}
