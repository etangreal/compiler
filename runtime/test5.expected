True
True
True
True
True
1
3
False
<object>
<object>
False
False
True
False
1
3
False
